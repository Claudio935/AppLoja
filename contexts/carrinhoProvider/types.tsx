export const types = {
    TYPE_ADD_CARRINHO:"TYPE_ADD_CARRINHO",
    TYPE_DELETE_CARRINHO:"TYPE_DELETE_CARRINHO"
}