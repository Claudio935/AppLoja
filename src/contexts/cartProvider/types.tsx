export enum types {
  TYPE_ADD_PRODUCT = "TYPE_ADD_PRODUCT",
  TYPE_DELETE_CART = "TYPE_DELETE_CART",
  TYPE_INCREMENT_CART = "TYPE_INCREMENT_CART",
  TYPE_DECREMENT_CART = "TYPE_DECREMENT_CART",
  TYPE_ADD_CART = "TYPE_ADD_CART",
  TYPE_CLEAN_CART = "TYPE_CLEAN_CART",
}