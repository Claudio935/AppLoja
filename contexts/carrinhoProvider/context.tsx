import React,{ createContext } from 'react'
import { Action, PropsCarrinho, PropsCarrinhoState } from '../../types/interfaces';









export const defaultValue: PropsCarrinhoState =  {
  carrinho:[
  ]}


export const CarrinhoContext = createContext<PropsCarrinhoState>(defaultValue)