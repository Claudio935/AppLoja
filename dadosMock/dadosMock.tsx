export const mockCarrinho = [{
    title: 'testMasculino',
    image: './img/testMasculino',
    price: 1.5,
    description: 'testandoMasculino',
    id: '1',
    category: "men's clothing"
  
  },
  {
    title: 'testFeminino',
    image: './img/testFeminino',
    price: 2,
    description: 'testandoFeminino',
    id: '2',
    category: "women's clothing"
  
  },
  {
    title: 'testEletronico',
    image: './img/testEletronico',
    price: 3,
    description: 'testandoEletronico',
    id: '3',
    category: "electronics"
  
  },
  
  {
    title: 'testJoia',
    image: './img/testJoia',
    price: 4,
    description: 'testandoJoia',
    id: '4',
    category: "jewelery"
  
  }
  ]