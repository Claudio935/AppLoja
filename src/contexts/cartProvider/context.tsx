import { createContext } from 'react'
import { type Action, type PropsCartState } from '../../types/types';


export const defaultValue: PropsCartState = {
  cart: []
}


export const CartContext = createContext<{
  store: PropsCartState;
  dispatch: React.Dispatch<Action>;
}>({
  store: defaultValue,
  dispatch: () => null
})