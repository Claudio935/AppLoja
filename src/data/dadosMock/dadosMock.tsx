export const mockCart = {
  cart: [{
    title: 'testMasculino',
    image: './img/testMasculino',
    price: 1.5,
    description: 'testandoMasculino',
    id: '1',
    category: "men's clothing",
    quantidade: 1,
  },
  {
    title: 'testFeminino',
    image: './img/testFeminino',
    price: 2,
    description: 'testandoFeminino',
    id: '2',
    category: "women's clothing",
    quantidade: 1,
  },
  {
    title: 'testEletronico',
    image: './img/testEletronico',
    price: 3,
    description: 'testandoEletronico',
    id: '3',
    category: "electronics",
    quantidade: 1,
  },

  {
    title: 'testJoia',
    image: './img/testJoia',
    price: 4,
    description: 'testandoJoia',
    id: '4',
    category: "jewelery",
    quantidade: 1,

  }
  ],
  dispatch: () => null
}

export const mockUseContext = {
  store: {
    cart:
      [
        {
          title: 'testMasculino',
          image: './img/testMasculino',
          price: 1.5,
          description: 'testandoMasculino',
          id: '102',
          category: "men's clothing",
          quantidade: 1,
        },
      ],
  },

  dispatch: jest.fn()
}