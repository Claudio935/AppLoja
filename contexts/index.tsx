import {useReducer} from "react"
import { CarrinhoContext,  defaultValue } from "./carrinhoProvider/context";
import { reducer } from "./carrinhoProvider/reducer";
import { PropsCarrinho } from "../types/interfaces";

interface Props {
   
    children: JSX.Element,
  };


export const CarrinhoProvider = ({children}:Props) =>{
      const [carrinho, setState] = useReducer(reducer, defaultValue)
 
    return<CarrinhoContext.Provider value={{carrinho: [...carrinho.carrinho], dispatch: setState}}>{children}</CarrinhoContext.Provider>
}