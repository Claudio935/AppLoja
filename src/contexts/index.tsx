import React, { useReducer, useEffect } from "react"
import AsyncStorage from '@react-native-async-storage/async-storage';
import { CartContext, defaultValue } from "./cartProvider/context";
import { reducer } from "./cartProvider/reducer";


type Props = {

  children: JSX.Element,
};


export const CartProvider: React.FC<Props> = ({ children }: Props) => {
  const [store, dispatch] = useReducer(reducer, defaultValue)


  useEffect(() => {
    // Pega o valor salvo
    AsyncStorage.getItem('store').then((value) => {

      if (!value) return
      // Atualiza meu contexto
      dispatch({ type: "RELOAD", payload: JSON.parse(value) })

    });

  }, []);

  useEffect(() => {

    // Salva o item localmente sendo a chave store
    AsyncStorage.setItem('store', JSON.stringify(store.cart));
  }, [store]);

  return <CartContext.Provider value={{ store, dispatch }}>{children}</CartContext.Provider>
}