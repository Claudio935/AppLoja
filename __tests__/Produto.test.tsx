import { NavigationContainer } from "@react-navigation/native";
import { render, screen, fireEvent } from "@testing-library/react-native";
import "@testing-library/jest-dom";
import React from "react";
import { Alert } from "react-native";
import renderer from 'react-test-renderer';
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import { RootStackParamList } from "../types/interfaces";
import Produto from "../views/produto/Produto";
import Carrinho from "../views/carrinho/Carrinho";
import { CarrinhoProvider } from "../contexts";
import { CarrinhoContext } from "../contexts/carrinhoProvider/context";
import { mockCarrinho } from "../dadosMock/dadosMock";


const mockUseContext = {
  carrinho: [
    mockCarrinho[0]
  ],
  dispatch: jest.fn()
}

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useRoute: () => ({
    params: mockCarrinho[0],
  }),
}));

jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  RN.Alert.alert = jest.fn();
  return RN;
});

describe("<Produto />", () => {
  it('renders correctly', () => {
    const Stack = createNativeStackNavigator<RootStackParamList>();

    const tree = renderer
      .create(<CarrinhoProvider>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Produto" component={Produto} />
            <Stack.Screen name="Carrinho" component={Carrinho} />
          </Stack.Navigator>
        </NavigationContainer>
      </CarrinhoProvider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should render and show components ", async () => {

    const Stack = createNativeStackNavigator<RootStackParamList>();


    render(
      <CarrinhoProvider>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Produto" component={Produto} />
            <Stack.Screen name="Carrinho" component={Carrinho} />
          </Stack.Navigator>
        </NavigationContainer>
      </CarrinhoProvider>

    );

    //testando a aparição do produto
    const imgProduto = screen.getByTestId('imageProduto');
    const titleProduto = screen.getByText(`${mockCarrinho[0].title}`)
    const descriptionProduto = screen.getByText(`${mockCarrinho[0].description}`);
    expect(imgProduto).toBeTruthy();
    expect(titleProduto).toBeTruthy();
    expect(descriptionProduto).toBeTruthy();
    expect(imgProduto.props.source.uri).toEqual(`${mockCarrinho[0].image}`);





  });
  it("changes quantity and price correctly.", async () => {

    const Stack = createNativeStackNavigator<RootStackParamList>();

    const useContextSpy = jest.spyOn(React, 'useContext');
    useContextSpy.mockReturnValue(mockUseContext);

    render(
      <CarrinhoContext.Provider value={mockUseContext}>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Produto" component={Produto} />
            <Stack.Screen name="Carrinho" component={Carrinho} />
          </Stack.Navigator>
        </NavigationContainer>
      </CarrinhoContext.Provider>

    );


    //testando adição de quantidade e a mudança do preço
    const textQuantidade = screen.getByTestId('textQuantidade');

    const buttonQuantidadeIncrement = screen.getByTestId("increment");
    const buttonQuantidadeDecrement = screen.getByTestId("decrement");
    const textButtonQuantidadeIncrement = screen.getByText(">");
    const textButtonQuantidadeDecrement = screen.getByText("<");

    expect(textButtonQuantidadeIncrement).toBeTruthy();
    expect(textButtonQuantidadeDecrement).toBeTruthy();

    expect(buttonQuantidadeIncrement).toBeTruthy();
    expect(buttonQuantidadeDecrement).toBeTruthy();

    fireEvent.press(buttonQuantidadeIncrement);
    fireEvent.press(buttonQuantidadeIncrement);

    expect(textQuantidade.props.children).toEqual(2);
    fireEvent.press(buttonQuantidadeDecrement);
    expect(textQuantidade.props.children).toEqual(1);




    const textPrice = screen.getByText(`R$ ${mockCarrinho[0].price}`);

    expect(textPrice).toBeTruthy();

    fireEvent.press(buttonQuantidadeDecrement);
    fireEvent.press(buttonQuantidadeDecrement);
    expect(textQuantidade.props.children).toEqual(0);





  });
  it("Test the quantity and price change action of an item on button click and show alert id", async () => {

    const Stack = createNativeStackNavigator<RootStackParamList>();

    const useContextSpy = jest.spyOn(React, 'useContext');
    useContextSpy.mockReturnValue(mockUseContext);

    render(
      <CarrinhoContext.Provider value={mockUseContext}>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Produto" component={Produto} />
            <Stack.Screen name="Carrinho" component={Carrinho} />
          </Stack.Navigator>
        </NavigationContainer>
      </CarrinhoContext.Provider>

    );


    const buttonQuantidadeIncrement = screen.getByTestId("increment");
    fireEvent.press(buttonQuantidadeIncrement);

    //testando o alert da quantidade
    const texButton = screen.getByText('Adicionar ao Carrinho');
    expect(texButton).toBeTruthy();
    const button = screen.getByTestId('carrinhoButton');

    fireEvent.press(button);

    expect(Alert.alert as jest.Mock).toHaveBeenCalledWith("Alerta",
      "Produto já adicionado?",
      [
        {
          text: "Cancelar",
          style: "cancel"
        },

      ]);




  });
  it("shows alert for lack of product quantity", async () => {

    const Stack = createNativeStackNavigator<RootStackParamList>();



    render(
      <CarrinhoProvider>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Produto" component={Produto} />
            <Stack.Screen name="Carrinho" component={Carrinho} />
          </Stack.Navigator>
        </NavigationContainer>
      </CarrinhoProvider>

    );



    const texButton = screen.getByText('Adicionar ao Carrinho');
    expect(texButton).toBeTruthy();
    const button = screen.getByTestId('carrinhoButton');

    fireEvent.press(button);


    expect(Alert.alert as jest.Mock).toHaveBeenCalledWith("Alerta",
      "Adicione uma quantidade de produto",
      [
        {
          text: "Cancelar",
          style: "cancel"
        },

      ]);

  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
});