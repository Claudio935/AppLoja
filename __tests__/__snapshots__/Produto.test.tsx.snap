// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Produto /> renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Produto"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <RCTScrollView>
            <View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "display": "flex",
                    "flexDirection": "column",
                    "justifyContent": "center",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "display": "flex",
                      "flexDirection": "column",
                      "height": 500,
                      "justifyContent": "space-between",
                      "marginTop": 10,
                      "width": "100%",
                    }
                  }
                >
                  <Image
                    source={
                      {
                        "uri": "./img/test",
                      }
                    }
                    style={
                      {
                        "alignItems": "center",
                        "display": "flex",
                        "height": 220,
                        "justifyContent": "center",
                        "width": 180,
                      }
                    }
                    testID="imageProduto"
                  />
                  <Text
                    style={
                      {
                        "color": "#000",
                        "fontSize": 15,
                        "fontWeight": "900",
                      }
                    }
                  >
                    testMasculino
                  </Text>
                  <Text
                    style={
                      {
                        "paddingHorizontal": 5,
                        "textAlign": "justify",
                      }
                    }
                  >
                    testandoMasculino
                  </Text>
                  <Text
                    style={
                      {
                        "color": "#b71919",
                        "fontSize": 14,
                        "fontWeight": "900",
                      }
                    }
                  >
                    R$ 0
                  </Text>
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "width": "100%",
                      }
                    }
                  >
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        {
                          "busy": undefined,
                          "checked": undefined,
                          "disabled": undefined,
                          "expanded": undefined,
                          "selected": undefined,
                        }
                      }
                      accessibilityValue={
                        {
                          "max": undefined,
                          "min": undefined,
                          "now": undefined,
                          "text": undefined,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        {
                          "opacity": 1,
                        }
                      }
                      testID="decrement"
                    >
                      <View
                        style={
                          [
                            {},
                          ]
                        }
                      >
                        <Text
                          style={
                            [
                              {
                                "color": "#007AFF",
                                "fontSize": 18,
                                "margin": 8,
                                "textAlign": "center",
                              },
                            ]
                          }
                        >
                          &lt;
                        </Text>
                      </View>
                    </View>
                    <Text
                      style={
                        {
                          "fontWeight": "900",
                          "marginHorizontal": 15,
                        }
                      }
                      testID="textQuantidade"
                    >
                      0
                    </Text>
                    <View
                      accessibilityRole="button"
                      accessibilityState={
                        {
                          "busy": undefined,
                          "checked": undefined,
                          "disabled": undefined,
                          "expanded": undefined,
                          "selected": undefined,
                        }
                      }
                      accessibilityValue={
                        {
                          "max": undefined,
                          "min": undefined,
                          "now": undefined,
                          "text": undefined,
                        }
                      }
                      accessible={true}
                      collapsable={false}
                      focusable={true}
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        {
                          "opacity": 1,
                        }
                      }
                      testID="increment"
                    >
                      <View
                        style={
                          [
                            {},
                          ]
                        }
                      >
                        <Text
                          style={
                            [
                              {
                                "color": "#007AFF",
                                "fontSize": 18,
                                "margin": 8,
                                "textAlign": "center",
                              },
                            ]
                          }
                        >
                          &gt;
                        </Text>
                      </View>
                    </View>
                  </View>
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "opacity": 1,
                      }
                    }
                    testID="carrinhoButton"
                  >
                    <View
                      style={
                        [
                          {},
                        ]
                      }
                    >
                      <Text
                        style={
                          [
                            {
                              "color": "#007AFF",
                              "fontSize": 18,
                              "margin": 8,
                              "textAlign": "center",
                            },
                          ]
                        }
                      >
                        Adicionar ao Carrinho
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </RCTScrollView>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
